contrib/nexusproto/serializer.go

```diff
@@ -0,0 +1,159 @@
+package nexusproto
+
+import (
+	"errors"
+	"fmt"
+	"mime"
+	"reflect"
+
+	"github.com/nexus-rpc/sdk-go/nexus"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/proto"
+)
+
+func messageFromAny(v any) (proto.Message, error) {
+	rv := reflect.ValueOf(v)
+	if rv.Kind() != reflect.Ptr {
+		return nil, fmt.Errorf("%w: type: %T", ErrInvalidProtoObject, v)
+	}
+	elem := rv.Type().Elem()
+
+	// v is a proto.Message or *proto.Message, the serializer supports either:
+	// 1. var d durationpb.Duration; s.Deserialize(&d)
+	// 2. var d *durationpb.Duration; s.Deserialize(&d)
+	// In the second case, we need to instantiate an empty proto struct:
+	if elem.Kind() == reflect.Ptr {
+		empty := reflect.New(elem.Elem())
+		rv.Elem().Set(empty)
+		rv = empty
+	}
+
+	msg, ok := rv.Interface().(proto.Message)
+	if !ok {
+		return nil, nexus.ErrSerializerIncompatible
+	}
+
+	return msg, nil
+}
+
+type protoJsonSerializer struct{}
+
+func (protoJsonSerializer) isValidContentType(contentType string) bool {
+	if contentType == "" {
+		return false
+	}
+	mediaType, params, err := mime.ParseMediaType(contentType)
+	return err == nil && mediaType == "application/json" && len(params) == 2 && params["format"] == "protobuf" && params["message-type"] != ""
+}
+
+func (s protoJsonSerializer) Deserialize(c *nexus.Content, v any) error {
+	if !s.isValidContentType(c.Header["type"]) {
+		return nexus.ErrSerializerIncompatible
+	}
+	msg, err := messageFromAny(v)
+	if err != nil {
+		return err
+	}
+	return protojson.Unmarshal(c.Data, msg)
+}
+
+func (protoJsonSerializer) Serialize(v any) (*nexus.Content, error) {
+	msg, ok := v.(proto.Message)
+	if !ok {
+		return nil, nexus.ErrSerializerIncompatible
+	}
+	data, err := protojson.Marshal(msg)
+	if err != nil {
+		return nil, err
+	}
+	messageType := string(msg.ProtoReflect().Descriptor().FullName())
+	return &nexus.Content{
+		Header: nexus.Header{
+			"type": mime.FormatMediaType("application/json", map[string]string{
+				"format":       "protobuf",
+				"message-type": messageType,
+			}),
```

Comment:
Can we update the Nexus spec with these details for supporting protobuf?

Reply:
Sure, I can do that.

Reply:
https://github.com/nexus-rpc/api/pull/18